### PR TITLE
191 create and airdrop eth account

### PR DESCRIPTION
**Pull request by rozhkovdmitrii**

### CODE REVIEW

File: proxy/plugin/solana_rest_api.py [hunk 1]
```diff
@@ -465,7 +470,8 @@ def eth_sendRawTransaction(self, rawTrx):
             return eth_signature
 
         except solana.rpc.api.SendTransactionError as err:
-            logger.debug("eth_sendRawTransaction solana.rpc.api.SendTransactionError:%s", err.result)
+            logs = "\n\t".join(err.result.get("data", {}).pop("logs", []))
```
Comment (rozhkovdmitrii) [inline proxy/plugin/solana_rest_api.py]: It needs to be changed **in place**

File: proxy/plugin/solana_rest_api.py [hunk 2]
```diff
@@ -465,7 +471,8 @@ def eth_sendRawTransaction(self, rawTrx):
...
+            logger.debug(f"eth_sendRawTransaction solana.rpc.api.SendTransactionError: {err.result}, {logs}")
```
Comment (rozhkovdmitrii) [inline proxy/plugin/solana_rest_api.py:474]: `"Failed to eth_getBalance, eror: {err.result}`

File: proxy/plugin/solana_rest_api_tools.py [hunk 3]
```diff
@@ -1041,20 +1040,26 @@ def call_signed_with_holder_acc(signer, client, ethTrx, perm_accs, trx_accs, ste
...
 
 
-def createEtherAccountTrx(client, ether, evm_loader_id, signer, code_acc=None):
-    if isinstance(ether, str):
-        if ether.startswith('0x'): ether = ether[2:]
-    else: ether = ether.hex()
-    (sol, nonce) = ether2program(ether, evm_loader_id, signer.public_key())
+def create_eth_account_trx(client: SolanaClient, signer: SolanaAccount, ether: EthereumAddress, evm_loader_id, code_acc=None) -> Tuple[Transaction, str, str]:
+
+    logger.debug(f"Signer: {signer}")
```
Comment (rozhkovdmitrii) [inline proxy/plugin/solana_rest_api_tools.py]: to be removed

File: proxy/plugin/solana_rest_api_tools.py [hunk 4]
```diff
@@ -1041,20 +1040,26 @@ def call_signed_with_holder_acc(signer, client, ethTrx, perm_accs, trx_accs, ste
...
+    sol, nonce = ether2program(ether, evm_loader_id, signer.public_key())
     associated_token = get_associated_token_address(PublicKey(sol), ETH_TOKEN_MINT_ID)
     logger.debug('createEtherAccount: {} {} => {}'.format(ether, nonce, sol))
```
Comment (rozhkovdmitrii) [inline proxy/plugin/solana_rest_api_tools.py]: Should be a single trace

**Comment by rozhkovdmitrii** [comment]
@s-medvedev кажется сначало нужно вот это смержить 
https://github.com/neonlabsorg/proxy-model.py/pull/282/files

**Comment by rozhkovdmitrii** [comment]
> @s-medvedev кажется сначало нужно вот это смержить https://github.com/neonlabsorg/proxy-model.py/pull/282/files

Обновил evm_loader ушла проблема

File: proxy/plugin/solana_rest_api.py [hunk 5]
```diff
@@ -474,6 +479,14 @@ def eth_sendRawTransaction(self, rawTrx):
             logger.debug("eth_sendRawTransaction type(err):%s, Exception:%s", type(err), err)
             raise
 
+    def _log_transaction_error(self, error: SolanaTrxError, logger):
+        copy.deep_copy(error)
```
Comment (rozhkovdmitrii) [inline proxy/plugin/solana_rest_api.py]: `deepcopy` raises error

File: proxy/plugin/solana_rest_api_tools.py [hunk 6]
```diff
@@ -1106,29 +1105,35 @@ def call_signed_with_holder_acc(signer, client, eth_trx, perm_accs, trx_accs, st
...
-        if ether.startswith('0x'): ether = ether[2:]
-    else: ether = ether.hex()
-    (sol, nonce) = ether2program(ether, evm_loader_id, signer.public_key())
-    associated_token = get_associated_token_address(PublicKey(sol), ETH_TOKEN_MINT_ID)
-    logger.debug('createEtherAccount: {} {} => {}'.format(ether, nonce, sol))
-    logger.debug('associatedTokenAccount: {}'.format(associated_token))
+def create_eth_account_trx(client: SolanaClient, signer: SolanaAccount, ether: EthereumAddress, evm_loader_id, code_acc=None) -> Tuple[Transaction, str, str]:
+
+    solana_address, nonce = ether2program(ether, evm_loader_id, signer.public_key())
+
```
Comment (rozhkovdmitrii) [inline proxy/plugin/solana_rest_api_tools.py]: too many spaces

File: proxy/plugin/solana_rest_api_tools.py [hunk 7]
```diff
@@ -1209,37 +1216,96 @@ def _getAccountData(client, account, expected_length, owner=None):
...
+                                                     program_id=TOKEN_PROGRAM_ID))
+    logger.debug(f"Token transfer from token: {owner_token_addr}, owned by: {owner_sol_addr}, to token: "
+                 f"{dest_token_account}, owned by: {dest_account} , value: {NEW_USER_AIRDROP_AMOUNT}")
+    trx.add(transfer_instruction)
+
+
+def create_token_and_airdrop_trx(client: SolanaClient, signer: SolanaAccount, eth_acc: EthereumAddress,
+                                 trx: Transaction, code_acc=None):
+    create_trx, solana_address, token_address = create_eth_account_trx(client, signer, eth_acc, evm_loader_id, code_acc)
+    logger.debug(f"Create token related to eth_acc: {eth_acc}, aka: {solana_address}, token_address: {token_address},"
```
Comment (rozhkovdmitrii) [inline proxy/plugin/solana_rest_api_tools.py]: extra logging

File: proxy/plugin/solana_rest_api_tools.py [hunk 8]
```diff
@@ -1106,29 +1105,33 @@ def call_signed_with_holder_acc(signer, client, eth_trx, perm_accs, trx_accs, st
...
 
-def createEtherAccountTrx(client, ether, evm_loader_id, signer, code_acc=None):
-    if isinstance(ether, str):
-        if ether.startswith('0x'): ether = ether[2:]
-    else: ether = ether.hex()
-    (sol, nonce) = ether2program(ether, evm_loader_id, signer.public_key())
-    associated_token = get_associated_token_address(PublicKey(sol), ETH_TOKEN_MINT_ID)
-    logger.debug('createEtherAccount: {} {} => {}'.format(ether, nonce, sol))
-    logger.debug('associatedTokenAccount: {}'.format(associated_token))
+def create_eth_account_trx(client: SolanaClient, signer: SolanaAccount, ether: EthereumAddress, evm_loader_id, code_acc=None) -> Tuple[Transaction, PublicKey]:
```
Comment (rozhkovdmitrii) [inline proxy/plugin/solana_rest_api_tools.py]: eth_address

File: proxy/testing/test_airdropping_eth_accounts.py [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+    def _compile_and_deploy_contract(self, contract_owner: LocalAccount, source: str) -> web3_eth.Contract:
+        compiled_sol = solcx.compile_source(source)
+        contract_id, contract_interface = compiled_sol.popitem()
+        contract = self._web3.eth.contract(abi=contract_interface['abi'], bytecode=contract_interface['bin'])
+        nonce = self._web3.eth.get_transaction_count(contract_owner.address)
+        chain_id = self._web3.eth.chain_id
+        trx_signed = self._web3.eth.account.sign_transaction(
+            dict(nonce=nonce, chainId=chain_id, gas=987654321, gasPrice=self._MINIMAL_GAS_PRICE, to='', value=0, data=contract.bytecode),
+            contract_owner.key
+        )
```
Comment (rozhkovdmitrii) [inline proxy/testing/test_airdropping_eth_accounts.py]: Shouldn't place the brace on the next line

File: proxy/plugin/solana_rest_api.py [hunk 10]
```diff
@@ -474,6 +479,14 @@ def eth_sendRawTransaction(self, rawTrx):
             logger.debug("eth_sendRawTransaction type(err):%s, Exception:%s", type(err), err)
             raise
 
+    def _log_transaction_error(self, error: SolanaTrxError, logger):
+        result = copy.deepcopy(error.result)
+        logs = result.get("data", {}).get("logs", [])
+        result.get("data", {}).update({"logs": ["\n\t" + log for log in logs]})
+        log_msg = str(result).replace("\\n\\t", "\n\t")
+        logger.debug(f"Got SendTransactionError: {log_msg}")
```
Comment (rozhkovdmitrii) [inline proxy/plugin/solana_rest_api.py]: logger.error of course!

File: proxy/common/utils.py [hunk 11]
```diff
@@ -208,3 +208,15 @@ def get_available_port() -> int:
         sock.bind(('', 0))
         _, port = sock.getsockname()
     return int(port)
+
+
+def get_from_dict(src: Dict, *path) -> Optional[Any]:
```
Comment (otselnik) [inline proxy/common/utils.py]: We are trying to touch the original proxy file less as possible. It is better to create your module for this.

File: proxy/plugin/solana_rest_api_tools.py [hunk 12]
```diff
@@ -934,26 +951,9 @@ def create_account_list_by_emulate(signer, client, eth_trx):
...
-                    mint=ETH_TOKEN_MINT_ID,
-                    owner=signer.public_key(),
-                    program_id=TOKEN_PROGRAM_ID,
-                    source=getTokenAddr(signer.public_key()),
-                )))
-                logger.debug("Token transfer to %s as ethereum 0x%s amount %s",
-                             get_associated_token_address(PublicKey(acc_desc["account"]), ETH_TOKEN_MINT_ID),
-                             acc_desc["address"],
-                             str(NEW_USER_AIRDROP_AMOUNT))
+            create_token_and_airdrop_trx(client, signer, EthereumAddress(address), trx, code_account)
```
Comment (otselnik) [inline proxy/plugin/solana_rest_api_tools.py]: The function shouldn't change the state of an object.
`create_token_and_airdrop_trx` should return Transaction or TransactionInstruction to be added to trx in this function.

Comment (otselnik) [reply proxy/plugin/solana_rest_api_tools.py]: Same applies to `add_airdrop_transfer_to_trx` and `create_token_and_airdrop_trx`.

**Review by otselnik — DISMISSED** [review]

File: proxy/testing/__init__.py [hunk 13]
```diff
@@ -8,3 +8,5 @@
     :copyright: (c) 2013-present by Abhinav Singh and contributors.
     :license: BSD, see LICENSE for more details.
 """
+import sys
```
Comment (s-medvedev) [inline proxy/testing/__init__.py:11]: This should be fixed in the testing script, not python tests. Bash script should prepare all necessary environments for launch tests.

File: proxy/plugin/solana_rest_api_tools.py [hunk 14]
```diff
@@ -1140,37 +1144,95 @@ def _getAccountData(client, account, expected_length, owner=None):
...
-    token_account = get_associated_token_address(PublicKey(account), ETH_TOKEN_MINT_ID)
 
-    balance = client.get_token_account_balance(token_account, commitment=Confirmed)
-    if 'error' in balance:
-        if NEW_USER_AIRDROP_AMOUNT > 0:
-            return NEW_USER_AIRDROP_AMOUNT * 1_000_000_000
-        else:
-            logger.debug("'error' in balance:")
-            return 0
+def get_transfer_instruction(owner_account: SolanaAccount, dest_token_account: PublicKey) -> TransactionInstruction:
```
Comment (s-medvedev) [inline proxy/plugin/solana_rest_api_tools.py]: I suggest this function rename to `make_transfer_instriction` (similarly to `make_partiac_call_instruction`)

File: proxy/plugin/solana_rest_api_tools.py [hunk 15]
```diff
@@ -1140,37 +1144,95 @@ def _getAccountData(client, account, expected_length, owner=None):
...
+    token_account = get_associated_token_address(PublicKey(token_owner_acc), ETH_TOKEN_MINT_ID)
+    rpc_response = client.get_token_account_balance(token_account, commitment=Confirmed)
+    error = rpc_response.get('error')
+    if error is None:
+        balance = get_from_dict(rpc_response, "result", "value", "amount")
+        if balance is None:
+            return None, f"Failed to get token balance from: {rpc_response}, by eth account:" \
+                         f" {eth_acc} aka: {token_account} at token: {ETH_TOKEN_MINT_ID}"
+        return int(balance), None
+    return None, error
```
Comment (s-medvedev) [inline proxy/plugin/solana_rest_api_tools.py]: Why do you pass the error message as result for this function? It seems to me that using exceptions for this is a more descriptive way.

File: proxy/plugin/solana_rest_api_tools.py [hunk 16]
```diff
@@ -867,26 +882,10 @@ def create_account_list_by_emulate(signer, client, eth_trx):
...
-                    owner=signer.public_key(),
-                    program_id=TOKEN_PROGRAM_ID,
-                    source=getTokenAddr(signer.public_key()),
-                )))
-                logger.debug("Token transfer to %s as ethereum 0x%s amount %s",
-                             get_associated_token_address(PublicKey(acc_desc["account"]), ETH_TOKEN_MINT_ID),
-                             acc_desc["address"],
-                             str(NEW_USER_AIRDROP_AMOUNT))
+            create_airdrop_trx = get_create_and_airdrop_trx(client, signer, EthereumAddress(address), code_account)
+            trx.add(create_airdrop_trx)
```
Comment (s-medvedev) [inline proxy/plugin/solana_rest_api_tools.py]: It turns out the unconditional addition of instructions for the transfer of tokens.

File: proxy/plugin/solana_rest_api_tools.py [hunk 17]
```diff
@@ -1171,57 +1176,64 @@ def get_transfer_instruction(owner_account: SolanaAccount, dest_token_account: P
...
+def extend_trx_with_create_and_airdrop(signer: SolanaAccount, eth_acc: EthereumAddress, code_acc=None, *, trx):
+    create_trx, token_address = make_create_eth_account_trx(signer, eth_acc, evm_loader_id, code_acc)
     trx.add(create_trx)
-    transfer_instruction = get_transfer_instruction(signer, token_address)
+    transfer_instruction = make_transfer_instruction(signer, token_address)
     trx.add(transfer_instruction)
-    return trx
 
 
 def create_token_and_airdrop(client: SolanaClient, signer: SolanaAccount, eth_acc: EthereumAddress):
```
Comment (s-medvedev) [inline proxy/plugin/solana_rest_api_tools.py]: Please, rename to `create_eth_account_and_airdrop`.

File: proxy/plugin/solana_rest_api_tools.py [hunk 18]
```diff
@@ -884,8 +889,8 @@ def create_account_list_by_emulate(signer, client, eth_trx):
                 trx.add(createAccountWithSeedTrx(signer.public_key(), signer.public_key(), seed, code_account_balance, code_size, PublicKey(evm_loader_id)))
                 code_account_writable = acc_desc["writable"]
 
-            create_airdrop_trx = get_create_and_airdrop_trx(client, signer, EthereumAddress(address), code_account)
-            trx.add(create_airdrop_trx)
+        if is_airdrop_allowed(client, signer, EthereumAddress(address)):
```
Comment (s-medvedev) [inline proxy/plugin/solana_rest_api_tools.py]: We should create eth_account in any case.
When NEW_USER_AIRDROP_AMOUNT is not zero we should add the instruction to airdrop tokens.

Comment (rozhkovdmitrii) [reply proxy/plugin/solana_rest_api_tools.py]: done

File: proxy/plugin/solana_rest_api_tools.py (hunk 17 above)
Comment (rozhkovdmitrii) [reply proxy/plugin/solana_rest_api_tools.py]: done

**Review by s-medvedev — APPROVED** [review]

**Pull request merged**